File: XiangXueApplication/app/src/main/java/com/ybx/xiangxue/kotlin/kt/core/TestOutIn.java
package com.ybx.xiangxue.kotlin.kt.core;

import java.util.ArrayList;
import java.util.List;

/**
 * @Author 55HAITAO
 * @Date 2020/7/6 4:26 PM
 */
class TestOutIn {

    FuClass fuClass = new FuClass();
    ZiClass ziClass = new ZiClass();

    void test01() {

        // 泛型读取模式
        // 只能取  不能改
        List<? extends FuClass> list = new ArrayList<ZiClass>();


//        list.add(fuClass)
        fuClass = list.get(0);
    }

    void test02() {
        List<? super ZiClass> list2 = new ArrayList<FuClass>();

        // 只能改 不能取
        list2.add(ziClass);
//        list2.add(fuClass);

//        ziClass =
//        list2.get(0);
    }
}

File: XiangXueApplication/app/src/main/java/com/ybx/xiangxue/kotlin/kt/core/TestOutIn.kt
package com.ybx.xiangxue.kotlin.kt.core

/**
 * @Author 55HAITAO
 * @Date 2020/7/6 4:25 PM
 */


fun test() {

    var ziClass: ZiClass = ZiClass()
    var fuClass: FuClass = FuClass()


    // 只取不改
    var list: MutableList<out FuClass> = ArrayList<ZiClass>()


    // 只改不取
    var list2: MutableList<in ZiClass> = ArrayList<FuClass>()
    list2.add(ziClass)
//    fuClass = list2[0]

}


// out ==== ? extent  只能取不能改
// in  ==== ? super   只能修改不能取

class TestOutIn<out T> {

    /*var list: MutableList<FuClass> = ArrayList()

    fun addData(data: T) {
        if (data is FuClass)
            list.add(data)
    }

    fun setData(data: T) {}*/


    fun getData(): T? {
        return null
    }

}


fun <T> test2(data: T) {


}